### ISSUE TITLE
GetCursorPosition requires chars in stdin

### ISSUE THREAD

**Issue by haf** [mono/mono#8875]
## Steps to Reproduce

1. See https://github.com/fsprojects/Paket/issues/3217#issue-325678788

## On which platforms did you notice this

[x] macOS
[ ] Linux
[ ] Windows

**Version Used**: 5.12.x

## Stacktrace

```
"Finalizer"
"<unnamed thread>"  at <unknown> <0xffffffff>
  at (wrapper managed-to-native) System.IO.MonoIO.Read (intptr,byte[],int,int,System.IO.MonoIOError&) [0x0000b] in <0f9df4881040473f9da7cf6c2e2cb8c3>:0
  at System.IO.MonoIO.Read (System.Runtime.InteropServices.SafeHandle,byte[],int,int,System.IO.MonoIOError&) [0x00010] in <0f9df4881040473f9da7cf6c2e2cb8c3>:0
  at System.IO.FileStream.ReadData (System.Runtime.InteropServices.SafeHandle,byte[],int,int) [0x00002] in <0f9df4881040473f9da7cf6c2e2cb8c3>:0
  at System.IO.FileStream.ReadInternal (byte[],int,int) [0x00026] in <0f9df4881040473f9da7cf6c2e2cb8c3>:0
  at System.IO.FileStream.Read (byte[],int,int) [0x000a1] in <0f9df4881040473f9da7cf6c2e2cb8c3>:0
  at System.IO.StreamReader.ReadBuffer () [0x000b3] in <0f9df4881040473f9da7cf6c2e2cb8c3>:0
  at System.IO.StreamReader.Read () [0x00021] in <0f9df4881040473f9da7cf6c2e2cb8c3>:0
  at System.TermInfoDriver.GetCursorPosition () [0x00048] in <0f9df4881040473f9da7cf6c2e2cb8c3>:0
  at System.TermInfoDriver.Init () [0x002dc] in <0f9df4881040473f9da7cf6c2e2cb8c3>:0
  at System.TermInfoDriver.get_WindowWidth () [0x00008] in <0f9df4881040473f9da7cf6c2e2cb8c3>:0
  at System.ConsoleDriver.get_WindowWidth () [0x00000] in <0f9df4881040473f9da7cf6c2e2cb8c3>:0
  at System.Console.get_WindowWidth () [0x00000] in <0f9df4881040473f9da7cf6c2e2cb8c3>:0
  at Argu.Utils.getDefaultCharacterWidth () [0x00000] in <74fc5bc3658141b2a1592235359daf3d>:0
  at Argu.ArgumentParser`1<Template_REF>..ctor (Microsoft.FSharp.Core.FSharpOption`1<string>,Microsoft.FSharp.Core.FSharpOption`1<string>,Microsoft.FSharp.Core.FSharpOption`1<int>,Microsoft.FSharp.Core.FSharpOption`1<Argu.IExiter>,Microsoft.FSharp.Core.FSharpOption`1<bool>) [0x0000c] in <74fc5bc3658141b2a1592235359daf3d>:0
  at Paket.Program.main () [0x00150] in <74fc5bc3658141b2a1592235359daf3d>:0
  at <StartupCode$paket>.$Paket.Program.main@ () [0x0000c] in <74fc5bc3658141b2a1592235359daf3d>:0
  at (wrapper runtime-invoke) object.runtime_invoke_void (object,intptr,intptr,intptr) [0x0004c] in <0f9df4881040473f9da7cf6c2e2cb8c3>:0
```

**Comment by migueldeicaza** [comment]
There are two possibilities, that the invoking code is a full pseudo terminal, but does not implement the complete terminal support and does not respond to the ESC + [ 6 n sequence with the position, or as detailed in the comments there might be a race condition as many processes try to read the input.

There are a few workaround for this:

* set the TERM environment variable to "dumb"
* redirect standard output, forcing the "this is not a terminal" mode
* do not call any of the Console APIs that require querying the terminal (WindowWidth in this case)

**Comment by matthid** [comment]
But the point is that this is a real terminal? @haf Can you clarify is this happens reproducible every time on the same place or always on different places? Also is there a different process involved or only a single `paket.exe` at this time?

I'm not an expert around this area but isn't this supposed to "just" work even when we have a wrapper process (like fake.exe in this case) - like nested shells are working as well? So multiple processes is not a problem only when they read std-input, correct?

@migueldeicaza Does closing mean that if this can be reproduced in a single binary by using the console-api's we still manually need to synchronize those or would that be considered a bug in mono (but you consider that scenario unlikely here)?

**Comment by haf** [comment]
Yeah, it's iTerm 2:

<img width="357" alt="screen shot 2018-05-31 at 08 03 25" src="https://user-images.githubusercontent.com/193115/40764386-1fc78456-64a9-11e8-903e-87064594b7f1.png">

Running zsh.

It happens every time, but it's unpredictable when and where; sometimes I have to press enter just once, other times I have to do it many times.

My eyes and gut feeling after reading the target helper code is that FAKE parallelises push operations up to 8 concurrent ones.

**Comment by lewurm** [comment]
this looks like a duplicate of https://bugzilla.xamarin.com/show_bug.cgi?id=26983 and was fixed via https://github.com/mono/mono/pull/8138 recently. 

do you mind testing with a master build? (have a look at the CI status, there are links to packages on every commit: https://github.com/mono/mono/commits/master )

**Comment by rolfbjarne** [comment]
This still happens:

 ```
(lldb) monobt
frame #0: 0x00007fff685a881e libsystem_kernel.dylib`read + 10
frame #1: 0x00000001042f60a9 mono`console_read(filehandle=0x00007fe2d0c73750, buffer=0x0000000108013250, numbytes=<unavailable>, bytesread=0x00007ffeeba89240) at w32file-unix.c:1630:7 [opt]
frame #2: 0x00000001042f0c82 mono`mono_w32file_read_or_write(read=<unavailable>, handle=<unavailable>, buffer=0x0000000108013250, numbytes=1024, bytesread=0x00007ffeeba89240, win32error=0x00007ffeeba893d8) at w32file-unix.c:2800:9 [opt]
frame #3: 0x000000010433232e mono`ves_icall_System_IO_MonoIO_Read(handle=0x0000000000000000, dest=<unavailable>, dest_offset=<unavailable>, count=1024, io_error=0x00007ffeeba893d8, error=<unavailable>) at w32file.c:543:11 [opt]
frame #4: 0x000000010434819f mono`ves_icall_System_IO_MonoIO_Read_raw(a0=0x0000000000000000, a1=0x00007ffeeba89350, a2=0, a3=1024, a4=0x00007ffeeba893d8) at icall-def.h:429:1 [opt]
frame #5: 0x1067df32a (wrapper managed-to-native) System.IO.MonoIO:Read (intptr,byte[],int,int,System.IO.MonoIOError&) [{0x7fe2d300a790} + 0x9a]  (0x1067df290 0x1067df3b5) [0x7fe2d0c11090 - MSBuild.dll]
frame #6: 0x0000000105f57fd7 mscorlib.dll.dylib`System_IO_FileStream_ReadData_System_Runtime_InteropServices_SafeHandle_byte___int_int + 87
frame #7: 0x0000000105f562ea mscorlib.dll.dylib`System_IO_FileStream_ReadInternal_byte___int_int + 154
frame #8: 0x0000000105f56128 mscorlib.dll.dylib`System_IO_FileStream_Read_byte___int_int + 264
frame #9: 0x0000000105f0ae26 mscorlib.dll.dylib`System_IO_StreamReader_ReadBuffer + 294
frame #10: 0x0000000105f09bdc mscorlib.dll.dylib`System_IO_StreamReader_Read + 76
frame #11: 0x0000000105c8b60e mscorlib.dll.dylib`System_TermInfoDriver_GetCursorPosition + 190

$ mono --version
Mono JIT compiler version 6.12.0.74 (2020-02/e9d3af508e4 Fri May 15 10:10:32 EDT 2020)
Copyright (C) 2002-2014 Novell, Inc, Xamarin Inc and Contributors. www.mono-project.com
	TLS:           
	SIGSEGV:       altstack
	Notification:  kqueue
	Architecture:  amd64
	Disabled:      none
	Misc:          softdebug 
	Interpreter:   yes
	LLVM:          yes(610)
	Suspend:       hybrid
	GC:            sgen (concurrent by default)
```

This is with macOS' Terminal.

The guess that it's a race condition with multiple processes reading stdin seems right, since it only happens when doing things in parallel (such as `make -j`).
